Source/Global/global.cpp

```diff
@@ -67,12 +70,15 @@ HRESULT singleton_access(
 
         if (!s_singleton)
         {
+            assert(!s_useCount);
             return E_HC_NOT_INITIALISED;
         }
 
-        singleton = std::move(s_singleton);
-        s_singleton.reset();
-
+        singleton = s_singleton;
+        if (!--s_useCount)
+        {
+            s_singleton.reset();
+        }
         return S_OK;
```

Comment:
you need to return an error if `s_useCount > 0` otherwise `cleanup_async` will try to run the rest of the cleanup